lib/Dialect/SV/Ops.cpp

```diff
@@ -90,9 +90,143 @@ static void printAlwaysAtPosEdgeOp(OpAsmPrinter &p, AlwaysAtPosEdgeOp op) {
+  p.printSymbolName(op.getName());
+  p.printRegion(op.body(),
+                /*printEntryBlockArgs=*/false,
+                /*printBlockTerminators=*/false);
+}
+
+//===----------------------------------------------------------------------===//
+// InterfaceSignalOp
+
+static ParseResult parseInterfaceSignalOp(OpAsmParser &parser,
+                                          OperationState &result) {
+  StringAttr nameAttr;
+  TypeAttr type;
+  if (parser.parseSymbolName(nameAttr, ::mlir::SymbolTable::getSymbolAttrName(),
+                             result.attributes) ||
+      parser.parseColon() ||
+      parser.parseAttribute(type, "type", result.attributes))
+    return failure();
+
+  return success();
+}
+
+static void printInterfaceSignalOp(OpAsmPrinter &p, InterfaceSignalOp op) {
+  p << op.getOperationName() << ' ';
+  p.printSymbolName(op.getName());
+  p << " : ";
+  p.printAttribute(op.getAttr("type"));
+}
+
+//===----------------------------------------------------------------------===//
+// InterfaceModPortOp
+
+static ParseResult parseInterfaceModPortOp(OpAsmParser &parser,
+                                           OperationState &result) {
+  StringAttr nameAttr;
+  if (parser.parseSymbolName(nameAttr, ::mlir::SymbolTable::getSymbolAttrName(),
+                             result.attributes) ||
+      parser.parseLParen())
+    return failure();
+
+  auto context = result.getContext();
+  SmallVector<Attribute, 8> ports;
+  while (true) {
+    NamedAttrList tmpAttrs;
+    StringAttr directionAttr;
+    StringAttr signalAttr;
+    if (parser.parseSymbolName(signalAttr, "signal", tmpAttrs) ||
+        parser.parseColon() ||
+        parser.parseAttribute(directionAttr, "direction", tmpAttrs))
+      break;
+
+    auto directionEnum =
+        symbolizeModPortDirectionAttr(directionAttr.getValue());
+
+    if (!directionEnum.hasValue()) {
+      return failure();
+    }
+
+    auto signalRef = FlatSymbolRefAttr::get(signalAttr.getValue(), context);
+
+    ports.push_back(ModPortStructAttr::get(directionAttr, signalRef, context));
+
+    if (!parser.parseOptionalRParen())
+      break;
+
+    parser.parseOptionalComma();
+  }
+
+  result.addAttribute("ports", ArrayAttr::get(ports, context));
+
+  return success();
+}
+
+static void printInterfaceModPortOp(OpAsmPrinter &p, InterfaceModPortOp op) {
+  p << op.getOperationName() << ' ';
+  p.printSymbolName(op.getName());
+  p << " (";
+
+  auto ports = op.ports();
+  for (size_t i = 0, e = ports.size(); i != e; ++i) {
+    auto port = ports[i].dyn_cast<ModPortStructAttr>();
+
+    p.printSymbolName(port.signal().getRootReference());
+    p << " : ";
+    p << port.direction();
+
+    if (i < e - 1) {
+      p << ", ";
+    }
+  }
+
+  p << ')';
+}
+
+static LogicalResult verifyInterfaceModPortOp(InterfaceModPortOp op) {
+  for (auto port : op.ports()) {
+    if (!port.dyn_cast<ModPortStructAttr>()) {
+      op.emitOpError("should have an array of ModPortStructAttr attributes");
```

Comment:
nit: You can return this directly. Diagnostics implicitly convert to failure()

Reply:
Also, please use 'isa' instead of 'dyn_cast' in a boolean context.

Reply:
Will do.